avocado_vt/plugins/vt_runner.py

```diff
@@ -0,0 +1,280 @@
+                    if error_message:
+                        raise exceptions.TestWarn("funcatexit failed with: %s" %
+                                                  error_message)
+
+                except:  # nopep8 Old-style exceptions are not inherited from Exception()
+                    stacktrace.log_exc_info(sys.exc_info(), 'avocado.test')
+                    if t_type is not None:
+                        error_message = funcatexit.run_exitfuncs(env, t_type)
+                        if error_message:
+                            self.log.error(error_message)
+                    try:
+                        env_process.postprocess_on_error(self, params, env)
+                    finally:
+                        self._safe_env_save(env)
+                    raise
+
+            finally:
+                # Post-process
+                try:
+                    try:
+                        params['test_passed'] = str(test_passed)
+                        env_process.postprocess(self, params, env)
+                    except:  # nopep8 Old-style exceptions are not inherited from Exception()
+
+                        stacktrace.log_exc_info(sys.exc_info(),
+                                                'avocado.test')
+                        if test_passed:
+                            raise
+                        self.log.error("Exception raised during postprocessing:"
+                                       " %s", sys.exc_info()[1])
+                finally:
+                    if self._safe_env_save(env) or params.get("env_cleanup",
+                                                              "no") == "yes":
+                        env.destroy()  # Force-clean as it can't be stored
+
+        except Exception as e:
+            try:
+                if params.get("abort_on_error") != "yes":
+                    raise
+                # Abort on error
+                logging.info("Aborting job (%s)", e)
+                if params.get("vm_type") == "qemu":
+                    for vm in env.get_all_vms():
+                        if vm.is_dead():
+                            continue
+                        logging.info("VM '%s' is alive.", vm.name)
+                        for m in vm.monitors:
+                            logging.info("It has a %s monitor unix socket at:"
+                                         " %s", m.protocol, m.filename)
+                        logging.info("The command line used to start it was:"
+                                     "\n%s", vm.make_create_command())
+                    raise exceptions.JobError("Abort requested (%s)" % e)
+            finally:
+                self.queue.put({'type': 'stderr',
+                                'log': traceback.format_exc().encode('utf-8')})
+                self.queue.put({'status': 'finished', 'result': 'error'})
+        self.queue.put({'status': 'finished', 'result': 'pass'})
+
+
+class VTTestRunner(nrunner.BaseRunner):
```

Comment:
Ok, I am getting a bit confused about the different runner functions here. But I guess this could be more generic than the current PR so I will be glad to hear your opinion. At least guessing from the `run` function here this seems more like a process spawner to me. I do notice however it inherits from `nrunner.BaseRunner` which is described as "Base interface for a Runner". *However*, there is an actual interface `avocado.core.plugin_interfaces.Runner` which doesn't have anything to do with this base runner. Then on top of this there is the `avocado.plugins.runner_nrunner.Runner` which does `run_suite` (more like a runner than a spawner) and actually inherits from the runner interface. So do you mind clarifying what are the roles of all these different runners? What is this class trying to accomplish and shouldn't runners be associated with the general scheduling (e.g. in a state machine) of all tests and not with a particular test's environment?

@clebergnu I am curious about your comments here too. It could be that I get confused by all the "runner" names used throught.

Reply:
The `avocado.core.plugin_interfaces.Runner` is the interface for running the core avocado. Right now we still have two architectures the legacy runner and the nrunner, that is why we need the `avocado.core.plugin_interfaces.Runner` interface and AFAIK this will be removed together with legacy runner. Legacy runner is `avocado.plugins.runner.TestRunner` and nrunner  is `avocado.plugins.runner_nrunner.Runner`. This core nrunner spawns the runner, based on Runnable `kind`, for each task. Those runners inherit from `nrunner.BaseRunner` and their work is to run tests and return results. The commit bb9980d7f25c5842483a10313f54732b18ea3fda from this PR is about creating such runner for vt-tests. I am not sure if this explains all your questions, so if you are still confused, please ask.

Reply:
Thanks a lot for the explanation! Then for the future (not related to this PR) I am just wondering if the "Runner of Runners" should simply be called "Scheduler" so that we could reserve the Runner class for running a test and returning the results as it seems more logical. The "next runner" should then simply be the test scheduler that makes sure to orchestrate all the runners through the task state machine. Just an idea for better clarity of the naming.

I don't think `avocado.core.plugin_interfaces.Runner` is planned for deprecation since this contradicts with the fact that `avocado.plugins.runner_nrunner.Runner` also inherits from it (as I mentioned above). It will probably remain a "scheduler interface" to be inherited from alternatives to the current next-runner (e.g. cases of more sophisticated scheduling strategies). @clebergnu Correct me if I am wrong but perhaps these comments are relevant to your own thoughts.

Reply:
> Thanks a lot for the explanation! Then for the future (not related to this PR) I am just wondering if the "Runner of Runners" should simply be called "Scheduler" so that we could reserve the Runner class for running a test and returning the results as it seems more logical. The "next runner" should then simply be the test scheduler that makes sure to orchestrate all the runners through the task state machine. Just an idea for better clarity of the naming.
> 
Hi @pevogam,

I see no issues in re-thinking the names of the key components.  It's true that there are some aspects here that resemble a scheduler.

> I don't think `avocado.core.plugin_interfaces.Runner` is planned for deprecation since this contradicts with the fact that `avocado.plugins.runner_nrunner.Runner` also inherits from it (as I mentioned above). It will probably remain a "scheduler interface" to be inherited from alternatives to the current next-runner (e.g. cases of more sophisticated scheduling strategies). @clebergnu Correct me if I am wrong but perhaps these comments are relevant to your own thoughts.

So by now it's clear that the `avocado.core.plugin_interfaces.Runner` is not specific to the nrunner, it's the interface that allows a *job* to run its tests (in a test suite).  We used to have implementations based on libvirt vms, remote hosts, etc, but the problem with those runners (now removed) is that they'd really run jobs, and not tests.  It was a conceptual and core problem because there was no standardized  way of executing just a test at a given place.

This is precisely what the "nrunner runners" do: the same thing that can be executed locally, say, `avocado-runner-exec-test task-run -i 1-/bin/true -u /bin/true -s 127.0.0.1:9090`, can be executed by a "ssh spawner" by running `ssh -L 9090:127.0.0.1:9090 avocado-runner-exec-test task-run -i 1-/bin/true -u /bin/true -s 127.0.0.1:9090`.

Now, that interface is still being used to cope with both the legacy runner and `runner_nrunner.py`.  Once the legacy runner is removed, that interface may prove to *not* be necessary anymore, given that the nrunner archicture allows, via spawners, tests to run be run at different "places".  Again, that is still TBD in my opinion.

Reply:
This explains a lot indeed. In that case I would recommend later on having having at least a generic scheduler-runner interface for other schedulers that use the task state machine in other ways than the current nrunner. In other words I recommend an interface with `run_suite` method that would allow custom decisions about the order and scheduling of tests.